### ISSUE TITLE
network_rules block support for Azure Function apps

### ISSUE THREAD

**Issue by chammik** [hashicorp/terraform-provider-azurerm#25109]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

As of now, when creating an Azure function app, TF doesn't seem to support setting this option.
![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/87511715/8eafe195-a87b-4b16-9a6c-bc2b8830be4e)

I'm looking to set this option during function app creation.

Tried using _ip_restriction_ block. While the IP restrictions are applied correctly, not being able to set the option shown in the screenshot during creation of the function app, means that the user would need to come back and set it manually, post creation. This is the current workaround in place, but causes issues during end to end automated deployments.

Could you please provide the capability to set the option shown in the screenshot. It seems Azure storage accounts get around the problem by using _network_rules_ block. Which doesn't seem to be supported on Function apps.

### New or Affected Resource(s)/Data Source(s)

azurerm_function_app

### Potential Terraform Configuration

```hcl
network_rules {
    default_action             = "Deny"
    ip_rules                   = ["100.0.0.1"]
    virtual_network_subnet_ids = [azurerm_subnet.example.id]
  }
```

**Comment by xiaxyi** [comment]
Thanks @chammik for raising this issue, let me check the feature and update you later.

**Comment by xiaxyi** [comment]
@chammik The option to only allow the selected public network access can be configured for the function app during app creation successfully from my side 
![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/92154856/1ec120b3-c3aa-4e39-87cf-2cdbc7c880a0)


Below is the example config that I used:
```
resource "azurerm_linux_function_app" "test-standard-network" {
  name                = "xiaixntest-LFA-standard-network"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  service_plan_id     = azurerm_service_plan.linux-standard.id

  storage_account_name       = azurerm_storage_account.test.name
  storage_account_access_key = azurerm_storage_account.test.primary_access_key

  site_config {
    ip_restriction {
      ip_address = "13.107.6.152/31,13.107.128.0/22"
      name       = "test-restriction"
      priority   = 123
      action     = "Allow"
      headers {
        x_azure_fdid      = ["55ce4ed1-4b06-4bf1-b40e-4638452104da"]
        x_fd_health_probe = ["1"]
        x_forwarded_for   = ["9.9.9.9/32", "2002::1234:abcd:ffff:c0a8:101/64"]
        x_forwarded_host  = ["example.com"]
      }
    }
  }
}
```

Would you mind sharing your config so I can try reproducing the issue from my side?

**Comment by chammik** [comment]
Thanks for the response @xiaxyi. The problem here, is that there's an Azure policy, that prevents us from setting the _public_network_access_enabled_ setting to True. Which is defaults to, when not specified.

Since opening this request, we've stumbled upon below, which does seem a suitable alternative.
https://github.com/hashicorp/terraform-provider-azurerm/pull/25131

We will test it when it's out. If that works, this can be closed.

**Comment by xiaxyi** [comment]
@chammik The feature should be released already, would you mind testing it and let us know the result?

**Comment by RSwarnkar** [comment]
Yes, I also want this feature. For Storage Account I see there is separate resource : `[azurerm_storage_account_network_rules](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_account_network_rules)`.
